### PR TITLE
Unique service exceptions

### PR DESCRIPTION
**Pull request by omgitsbillryan**
Related to Item # 5 in this list:
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/9997#issuecomment-382917392

Sentry contains a large list of `BackendServiceException` entries that all have the same (or very similar) title but represent different errors.  This PR tries to define a unique exception within each `client`/`service` class so as to make Sentry more readable.

EDIT: I think this is also: https://github.com/department-of-veterans-affairs/vets-api/pull/1905

### CODE REVIEW

**Comment by omgitsbillryan** [comment]
So far I've only implemented this on `letters` as a proof of concept, but would like to extend this to all the [offenders found in this list](http://sentry.vetsgov-internal/vets-gov/platform-api-production/?query=is%3Aunresolved+BackendServiceException).

For the lazy, I'll list the top 5 most recent offenders listed in Sentry at the time of writing this comment:
- `app/services/mhv_accounts_service.rb` in `upgrade` at line 46
- `app/services/mhv_accounts_service.rb` in `create` at line 28
- `lib/evss/service.rb` in `raise_backend_exception` at line 45
- `lib/common/client/concerns/mhv_session_based_client.rb` in `authenticate` at line 16
- `lib/common/client/concerns/mhv_session_based_client.rb` in `authenticate` at line 16

File: lib/common/exceptions/external/backend_service_exception.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@ module Exceptions
     # you must define the minor code in the locales file and call this class from
     # raise_error middleware.
     class BackendServiceException < BaseError
-      attr_reader :response_values, :original_status, :original_body
+      attr_reader :response_values, :original_status, :original_body, :key
```
Comment (omgitsbillryan) [inline lib/common/exceptions/external/backend_service_exception.rb]: This was added so that `BackendServiceException` may be easily converted into it's service-specific form, [seen here](https://github.com/department-of-veterans-affairs/vets-api/pull/1905/files#diff-788ea7fbcaa41ab6482eb1f4d6eb3342R126)

**Comment by jkassemi** [comment]
@saneshark - great comments - appreciate seeing this kind of discussion in reviews. 

I totally get the desire to avoid the additional complexity that this introduces. 

My experiences on the project with sentry have been primarily during incident response, and generally due to partner service failures (either outages or issues with a new release). When those pop up, it's useful to quickly see which service or services are failing to determine if it's a common dependency that's having issues (BGS), or if there's a problem with a data center. We have some additional monitoring at the network level, but nothing beats checking out the exceptions the application directly raises. This change would provide that information pretty quickly, and I find that valuable and worth the additional complexity.

The second issue we struggle with is clear ownership of exceptions in sentry. In some cases, the exception is gracefully handled by our system, but we're still seeing an exception logged. I suspect that's one component in #4 from vets.gov-team#9997. I suspect this leads to conditions where we're not viewing `BackendServiceException` as critical. Further classification with a distinct title will help us better assign ownership of resolving the exceptions. One alternative to this (and probably a better approach) would be for each team to scan through all backend exceptions and close out any that are relevant to them. I'd probably start by selecting a past point in time and clearing the plate before then. This might be a parallel effort, and maybe we can eventually refactor this change away. The other option would be to set up additional isolation and log exceptions to separate Sentry projects. We've had discussions on the cons of that approach and they were pretty overwhelming - at least for where we are with the project right now.

It's fine to swallow errors that we're handling gracefully - but I don't think that's true with the majority of the exceptions that we're catching from partner services. I think we're still not getting maintenance windows in place consistently enough for planned outages, so there may be some additional work we can do to reduce the instances of `BackendServiceException` outside of this. 

I'm not a big fan of the `capture_message` option - particularly with dynamic messages. Certain services may be invoked from multiple locations, and I can envision cases where a specific caller may be invoking the service incorrectly. The error would be hard to track down without that traceback. The traceback could be included in additional context on the exception, but then we're going to muck around with Sentry's grouping mechanisms, which could lead to additional problems. I still can't figure out where http://sentry.vetsgov-internal/vets-gov/platform-api-production/issues/37302/ is raised. 

tldr; I see a lot of value in this approach despite the additional complexity - once we've cleared out our current set of exceptions currently tracked in Sentry, are practicing exception grooming consistently, and more gracefully handle these outages, can we revisit?

**Comment by omgitsbillryan** [comment]
- I agree that this adds "cruft code", but I wouldn't call it _complexity_. The code is blunt, and straight forward to me (granted I wrote it 😛). When another dev see's `RxServiceException` in Sentry, they'll `ctrl+f` through our code and easily figure out what's going on.

> I still can't figure out where http://sentry.vetsgov-internal/vets-gov/platform-api-production/issues/37302/ is raised

This made me 😆 and 😢 a little.  I'm weary of trying to be "too cute", for this reason.  It's more important to be easily auditable than to be perfectly DRY (to a point obviously).

- [Using `tags_context`](https://github.com/getsentry/raven-ruby#context) may be a more elegant solution than what I'm currently doing [converting `BackendServiceException` into the relevant service-level exception](https://github.com/department-of-veterans-affairs/vets-api/pull/1905/files#diff-788ea7fbcaa41ab6482eb1f4d6eb3342R64).  The problem with a tag is that not everyone would know it exists, whereas a title is a bit more idiot-proof.

- The ideal solution would be for us to go through all of our Sentry `error`'s and _address_ them.  This has not happened in practice, thus making a PR like this necessary because we're starting to lose auditability.

**Comment by jkassemi** [comment]
Potential con from the tags_context side before we address #3 from https://github.com/department-of-veterans-affairs/vets.gov-team/issues/9997#issuecomment-382917392 - context will be thread specific during the request, so it's possible (though I don't think we have any instances of this right now) that we'll end up adding invalid context in cases where we raise multiple errors.

**Comment by omgitsbillryan** [comment]
> so it's possible (though I don't think we have any instances of this right now) that we'll end up adding invalid context in cases where we raise multiple errors.

^^ Very good point

**Comment by kreek** [comment]
> I still can't figure out where http://sentry.vetsgov-internal/vets-gov/platform-api-production/issues/37302/ is raised

I'm pretty sure this is the source
https://github.com/department-of-veterans-affairs/vets-api/blob/45fbdbfb52bc936a5ec142c30b65144f37b455ec/app/controllers/v0/in_progress_forms_controller.rb#L22

The other source of 404s in sentry is when the base client rescues `Faraday::ResourceNotFound` and raises `Common::Client::Errors::ClientError`. Maybe that should be `Common::Client::Errors::NotFound` and then add that to sentry's ignore list as it's not unexpected? 

I agree with @saneshark on using the "RX", "SM" pattern, in the past I've used something similar but with dot notation like `app.service.record_not_found` which mapped to a YAML file similar to `exceptions.en.yml`. For EVSS we use that pattern, not a big change but less obscure than a single code (I know what 'VA900' is but I don't think new devs do 😄). I have not idea if you could leverage hierarchal dot notation to filter errors into buckets or apply tags, but it'd be cool if you could.

Lastly I think some of the double logging could also be reduced by catching things earlier in middleware and raising the error you want rather than it hopping through a few rescues and re-raises.

**Comment by saneshark** [comment]
@jkassemi @omgitsbillryan with respect to the criticism of capture_message not including the backtrace, it looks like this PR introduced the ability to add backtrace to capture_message in the exact same way as it is done for capture_exception: getsentry/raven-ruby#66 so there shouldn't be a problem with grouping anymore I don't think.

I think the real problem I have with this approach of introducing new error classes is that it doesn't really help in grouping trends as effectively as we would want. There are about 15-20 different errors defined in exception.en.yml for RX alone. Another 15-20 for SM. It would be much more valuable to have them grouped by exception message than by the RxBackendServiceException.

If we go do this route, can we at least come up with a way of ensuring that the error class is always consistent? `"#{ServiceNameMneumonic}ServiceException"`

File: lib/bb/client.rb [hunk 2]
```diff
@@ -9,13 +9,15 @@
...
   # Core class responsible for api interface operations
+  class BBServiceException < Common::Exceptions::BackendServiceException; end
   class Client < Common::Client::Base
     include Common::Client::MHVSessionBasedClient
     include Common::Client::StreamingClient
     include SentryLogging
 
     configuration BB::Configuration
     client_session Rx::ClientSession
+    use_service_exception BBServiceException
```
Comment (saneshark) [inline lib/bb/client.rb]: Since Common::Configuration already defines a mnemonic for the service name, i think you could possible remove the need for "use_service_exception" and just define service_exception on Common::Configuration, it can be automatically generated using minimal meta programming and will help to ensure some consistency.

File: lib/common/client/base.rb [hunk 3]
```diff
@@ -14,10 +14,20 @@ class Base
...
         end
+
+        def use_service_exception(exception)
+          unless exception.ancestors.include?(Common::Exceptions::BackendServiceException)
+            raise "[#{exception}] must inherit from BackendServiceException"
+          end
+          @service_exception = exception
+        end
       end
+      delegate :service_exception, to: 'self.class'
```
Comment (saneshark) [inline lib/common/client/base.rb]: I think a better home for this would be in Common::Configuration, since that is where we typically define configuration options. Additionally since the service_name is there, it would be ideal if this error class could live there as well. I'm not so sure the `unless` block will be necessary, it's trivial to generate a class based on the service_name that is already defined. The class will automatically be generated when Common::Configuration is initialized, and since Common::Configuration is a singleton, it's really a perfect place to this (no risk of initializing the constant twice).

File: lib/common/exceptions/external/backend_service_exception.rb (hunk 1 above)
Comment (saneshark) [reply lib/common/exceptions/external/backend_service_exception.rb]: can you elaborate on this a bit? I'm not sure I understand. Though I think adding key as an attribute could be useful.

File: lib/common/client/base.rb (hunk 3 above)
Comment (saneshark) [reply lib/common/client/base.rb]: something like:

```
Object.const_set("#{service_name}ServiceException", Class.new(BackendServiceException))
```

would dynamically generate the class based on service_name when configuration is initialized, and then Common::Client would define service_exception as always being:

"#{self.config.service_name}ServiceException".constantize || BackendServiceException

**Comment by omgitsbillryan** [comment]
> I think the real problem I have with this approach of introducing new error classes is that it doesn't really help in grouping trends as effectively as we would want. There are about 15-20 different errors defined in exception.en.yml for RX alone. Another 15-20 for SM. 

This is a good area for debate.  If we allow top-level Sentry entries be a combo of <SERVICE_NAME + ERROR_ CODE>, we'll end up with more noise.  The number of services we have is growing, and we only have one shared space for Sentry, so I don't think this approach scales.  Most devs know what `RxService` is, but a much smaller percentage know or care what `RxService-113` is. This approach creates a pollution problem.

I do recognize however, that this approach _limits_ visibility and is possible for errors to get buried in their respective bucket and never looked at.  For this, I think we'll have to put more care into how we tag things.

**Comment by omgitsbillryan** [comment]
> In other words, all of the RxServiceExceptions are not going to automatically get grouped into one category anymore than BackendServiceException currently does.

Very good point.  Really the only benefit this PR provides is the service name in the exception title.  [Grouping & rollup is done by stack trace alone if it exists according to their docs](https://docs.sentry.io/learn/rollups/), so we may just have _that many_ unique stack traces.  This will have to be a separate investigation.

File: lib/common/exceptions/external/backend_service_exception.rb (hunk 1 above)
Comment (saneshark) [reply lib/common/exceptions/external/backend_service_exception.rb]: I think it would work for initialization, but not work for reading it, but it's fine. doesn't hurt to make it accessible.

**Comment by omgitsbillryan** [comment]
This is becoming a bikeshedding activity. Closing due to high effort + high risk + low impact.  Our current error handling doesn't easily lend itself to the desired behavior of unique log messages per external `service` / `client` and trying to force this could have unintended consequences for our normal error responses.  Improved logging organization is only a cure to the symptom and not the real problem, which is, log entries in Sentries that go unattended to and eventually stale.

**Comment by jkassemi** [comment]
I'm disappointed to see this closed out. I feel like it's a valuable interim step that will aid parallel efforts toward improving ownership of issues in Sentry. I feel like this has benefit as it stands; if we feel that there are considerable risks that we're not anticipating, then I think further discussion is warranted, but I'd like to see us roll forward with this. 

Let's go for an RFC with a limited feedback window for changes like this in the future. Via PR to vets.gov-team.
